io.openems.edge.controller.ess.timeofusetariff/src/io/openems/edge/controller/ess/timeofusetariff/TimeOfUseTariffControllerImpl.java

```diff
@@ -0,0 +1,469 @@
+			// Sets the charge or discharge of the system based on the mode.
+			this.setChargeOrDischarge();
+			return;
+		}
+
+		// Prediction values
+		final var predictionProduction = this.predictorManager.get24HoursPrediction(SUM_PRODUCTION) //
+				.getValues();
+		final var predictionConsumption = this.predictorManager.get24HoursPrediction(SUM_CONSUMPTION) //
+				.getValues();
+
+		// Prices contains the price values and the time it is retrieved.
+		final var prices = this.timeOfUseTariff.getPrices();
+		this._setTotalQuarterlyPricesChannel(prices.getValues().length);
+
+		// Ess information.
+		final var netEssCapacity = this.ess.getCapacity().getOrError();
+		final var soc = this.ess.getSoc().getOrError();
+
+		// Calculate available energy using "netCapacity" and "soc".
+		var currentAvailableEnergy = netEssCapacity /* [Wh] */ / 100 * soc;
+		this.channel(TimeOfUseTariffController.ChannelId.AVAILABLE_CAPACITY).setNextValue(currentAvailableEnergy);
+
+		// Calculate the net usable energy of the battery.
+		final var limitEnergy = this.getLimitEnergy(netEssCapacity);
+		final var netUsableEnergy = TypeUtils.max(0, netEssCapacity - limitEnergy);
+
+		// Calculate current usable energy [Wh] in the battery.
+		currentAvailableEnergy = TypeUtils.max(0, currentAvailableEnergy - limitEnergy);
+		this.channel(TimeOfUseTariffController.ChannelId.USABLE_CAPACITY).setNextValue(currentAvailableEnergy);
+
+		// Power Values for scheduling battery for individual periods.
+		var power = this.ess.getPower();
+		var dischargePower = power.getMaxPower(this.ess, Phase.ALL, Pwr.ACTIVE);
+		var chargePower = power.getMinPower(this.ess, Phase.ALL, Pwr.ACTIVE);
+
+		// If both values are 0, its likely that components are not yet activated
+		// completely.
+		// TODO: Handle cases where max and min power are very low in beginning.
+		if (dischargePower == 0 && chargePower == 0) {
+			return;
+		}
+
+		// Power to Energy.
+		var dischargeEnergy = dischargePower / PERIODS_PER_HOUR;
+		var chargeEnergy = chargePower / PERIODS_PER_HOUR;
+		var allowedChargeEnergyFromGrid = this.config.maxPower() / PERIODS_PER_HOUR;
+
+		// Initialize Schedule
+		this.schedule = new Schedule(this.config.controlMode(), netUsableEnergy, currentAvailableEnergy,
+				dischargeEnergy, chargeEnergy, prices.getValues(), predictionConsumption, predictionProduction,
+				allowedChargeEnergyFromGrid);
+
+		// Generate Final schedule.
+		this.schedule.createSchedule();
+
+		// Update next quarter.
+		this.nextQuarter = now.plus(TIME_PER_PERIOD);
+
+		// Sets the charge or discharge of the system based on the mode.
+		this.setChargeOrDischarge();
+	}
+
+	/**
+	 * Force charges or delays the discharge if the schedule is set for current
+	 * period.
+	 * 
+	 * @throws OpenemsNamedException on error.
+	 */
+	private void setChargeOrDischarge() throws OpenemsNamedException {
+		if (this.schedule.periods.isEmpty()) {
+			return;
```

Comment:
If this method throws an Exception or returns here, the channels will not get set.